d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,122 @@
+package com.linkedin.d2.balancer.dualread;
+
+import com.linkedin.d2.balancer.properties.UriProperties;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UriPropertiesDualReadMonitor
+{
+  private static final Logger LOG = LoggerFactory.getLogger(UriPropertiesDualReadMonitor.class);
+
+  private final Map<String, Cluster> _clusters = new HashMap<>();
+  private double _totalUris = 0;
+  private double _matchedUris = 0;
+  private final DualReadLoadBalancerJmx _dualReadLoadBalancerJmx;
+
+  public UriPropertiesDualReadMonitor(DualReadLoadBalancerJmx dualReadLoadBalancerJmx)
+  {
+    _dualReadLoadBalancerJmx = dualReadLoadBalancerJmx;
+  }
+
+  public void reportData(String clusterName, UriProperties property, boolean fromNewLb)
+  {
+    Cluster cluster = _clusters.computeIfAbsent(clusterName, k -> new Cluster());
+
+    if (fromNewLb)
+    {
+      cluster._newLb = property;
+    }
+    else
+    {
+      cluster._oldLb = property;
+    }
+
+
+    _totalUris -= cluster._uris;
+    _matchedUris -= cluster._matched;
+
+    if (cluster._oldLb == null && cluster._newLb == null)
+    {
+      _clusters.remove(clusterName);
+      return;
+    }
+
+    cluster._matched = 0;
+
+    if (cluster._oldLb == null || cluster._newLb == null)
+    {
+      cluster._uris = (cluster._oldLb == null) ? cluster._oldLb.Uris().size() : cluster._newLb.Uris().size();
+      _totalUris += cluster._uris;
+      return;
+    }
+
+    cluster._uris = cluster._oldLb.Uris().size();
+    Set<URI> newLbUris = new HashSet<>(cluster._newLb.Uris());
+
+    for (URI uri : cluster._oldLb.Uris())
+    {
+      if (!newLbUris.remove(uri))
+      {
+        continue;
+      }
+
+      if (compareURI(uri, cluster._oldLb, cluster._newLb))
+      {
+        cluster._matched++;
+      }
+    }
+
+    cluster._uris += newLbUris.size();
+
+    _totalUris += cluster._uris;
+    _matchedUris += cluster._matched;
+    _dualReadLoadBalancerJmx.setUriPropertiesSimilarity(_matchedUris / _totalUris);
```

Comment:
at the end, no matter it's matched or not, we could log both oldLb and newLb properties to debug level for debugging when needed, similar to what we had "LOG.debug("Current entries of {} for {}: {}", propertyClassName, propertyName, entriesLogMsg);", where entriesLogMsg is 
"String.format("\nOld LB: %s\nNew LB: %s", oldLb, newLb);"

Reply:
Yeah makes sense